### PR TITLE
Add multihead training capability

### PR DESCRIPTION
**Pull request by sovrasov**

### CODE REVIEW

**Comment by sovrasov** [comment]
@kprokofi hierarchical classification still needs careful evaluation, but at least this PR shouldn't break the existing functionality.

**Comment by kprokofi** [comment]
@sovrasov , please, check https://github.com/openvinotoolkit/training_extensions/pull/888 if it is not break anything

**Comment by sovrasov** [comment]
888 breaks the initial label schema generation. Fixed it.

File: torchreid/data/datasets/dataset.py [hunk 1]
```diff
@@ -104,7 +113,10 @@ def parse_data(data):
         for record in data:
             label = record[1]
             if isinstance(label, (list, tuple)):
-                ids.update(set(label))
+                if len(label) and isinstance(label[0], (tuple, list)):
+                    ids.update([1])
```
Comment (kprokofi) [inline torchreid/data/datasets/dataset.py]: why [1]?

File: torchreid/data/datasets/image/classification.py [hunk 2]
```diff
@@ -257,3 +263,97 @@ def load_annotation(annot_path, data_dir):
...
+                    class_to_idx[c] = (i, j) # group idx and idx inside group
+                    num_single_label_classes += 1
+
+            # other labels are in multilabel group
+            for j, g in enumerate(single_label_groups):
+                class_to_idx[g[0]] = (len(exclusive_groups), j)
+
+            mixed_cls_heads_info = {
+                                    'num_multiclass_heads': len(exclusive_groups),
+                                    'num_multilabel_classes': len(single_label_groups),
```
Comment (kprokofi) [inline torchreid/data/datasets/image/classification.py:330]: Should `num_multilabel_classes` represent the number of `exclusive_groups` either? For example, we have 3 exclusive groups with some amount of labels and 4 other `mutual` labels. So, we need to define 7 multilabel classes (3 classes will be from exclusive groups and 4 other classes)

Comment (sovrasov) [reply torchreid/data/datasets/image/classification.py:330]: Each exclusive group is handled by a softmax-like loss, while all other classes are considered as multilabel subtask. `num_multilabel_classes` is the number of labels in this subtask. At the same time some of the exclusive groups could be mutually exclusive between themselves, so we can't add them to the multilabel subtask. This information is not passed to d-o-r, otherwise we have to support a description of hierarchy there.

File: torchreid/data/datasets/dataset.py (hunk 1 above)
Comment (sovrasov) [reply torchreid/data/datasets/dataset.py]: It was a temporal workaround. Forgot to get rid of it.

**Comment by kprokofi** [comment]
There are some conflicts after pylint merging

**Comment by kprokofi** [comment]
Have the tests for classification and multi-label classification passed?

**Comment by sovrasov** [comment]
> Have the tests for classification and multi-label classification passed?

These usecases are runnable, metrics are growing during training.

**Comment by kprokofi** [comment]
I've checked that after these changes efficient net still can achieve high accuracy

**Review by kprokofi — APPROVED** [review]
